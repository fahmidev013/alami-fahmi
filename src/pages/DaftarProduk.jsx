import PropTypes from "prop-types";
import React from "react";
import { useSelector, useDispatch } from "react-redux";
import TableComponent from "../components/TableComponent";
import {
  selectProductList,
  getProducts,
  getProductsByKeyword,
} from "../redux/reducer/produkReducer";

export const DaftarProduk = () => {
  const products = useSelector(selectProductList);
  const dispatch = useDispatch();

  React.useEffect(() => {
    dispatch(getProductsByKeyword(''));
  }, []);

  const submit = (event) => {
    event.preventDefault();
    if (document.getElementById('sellerId').value !== ''){
        dispatch(getProducts(document.getElementById('sellerId').value));
    } else {
        dispatch(getProductsByKeyword(''));
    }
    
  };

  const reset = (event) => {
    event.preventDefault();
    document.getElementById('sellerId').value = '';
    dispatch(getProductsByKeyword(''));
  }

  return (
      <div className="site-layout-content">
        <form className='inputForm' onSubmit={submit}>
          <input
            type="text"
            id="sellerId"
            placeholder="Ketik ID Seller"
            name="sellerId"
          />
          <button type="submit">Filter</button>&nbsp;&nbsp;&nbsp;
          <button onClick={reset}>Reset</button>
        </form>
        {products.length > 0 && (
          <div style={{ overflow: "scroll", maxHeight: "65vh" }}>
            <TableComponent data={products} />
          </div>
        )}
      </div>
  );
};

DaftarProduk.propTypes = {
  dataReducer: PropTypes.array,
};

export default DaftarProduk;
