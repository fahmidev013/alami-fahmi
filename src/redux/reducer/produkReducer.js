import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import axios from "axios";

const initialState = {
  list: [],
  loading: false,
};

export const getProducts = createAsyncThunk(
  "products/getProducts",
  async (id) => {
      try {
        const data = await axios
        .get(`listProductBySellerId?seller_id=${id}`)
        .then((response) => response.data)
        .then((result) => result.data.map((item) => ({ key: item.id, ...item })));
        console.log(JSON.stringify(data))
        return data;
      } catch (error) {
        return [];  
      }
  }
);

export const getProductsByKeyword = createAsyncThunk(
    "products/getProductsByKeyword",
    async (keyword) => {
        try {
          const data = await axios
          .get(`searchProductByKeyword?keyword=${keyword}`)
          .then((response) => response.data)
          .then((result) => result.data.map((item) => ({ key: item.id, ...item })));
          return data;
        } catch (error) {
          return [];  
        }
    }
  );

export const productReducer = createSlice({
  name: "products",
  initialState,
  reducers: {
    reset: (state) => {
      state.list = [];
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getProducts.pending, (state) => {
        state.loading = true;
      })
      .addCase(getProducts.fulfilled, (state, action) => {
        state.loading = false;
        state.list = action.payload;
      })
      .addCase(getProducts.rejected, (state) => {
        state.loading = false;
      })
      .addCase(getProductsByKeyword.pending, (state) => {
        state.loading = true;
      })
      .addCase(getProductsByKeyword.fulfilled, (state, action) => {
        state.loading = false;
        state.list = action.payload;
      })
      .addCase(getProductsByKeyword.rejected, (state) => {
        state.loading = false;
      });
  },
});

export const { reset } = productReducer.actions;

export const selectProductList = (state) => state.products.list;

export default productReducer.reducer;
