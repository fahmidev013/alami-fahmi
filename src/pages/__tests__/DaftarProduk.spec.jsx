import React from "react";
import DaftarProduk from "../DaftarProduk";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import produkReducer, {
  reset,
  getProducts,
} from "../../redux/reducer/produkReducer";
import { configureStore } from "@reduxjs/toolkit";
import axios from "axios";
import * as redux from "react-redux";

const data = [
  {
    key: 25,
    id: 25,
    nama: "Alpukat",
    satuan: "kg",
    hargaSatuan: 35000,
    sellerId: 65,
    deskripsi: "Alpukat Mentega",
  },
  {
    key: 31,
    id: 31,
    nama: "Balsem",
    satuan: "botol",
    hargaSatuan: 2000,
    sellerId: 65,
    deskripsi: "Balsem Oles",
  },
];

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn().mockReturnValue(data),
  useDispatch: jest.fn(),
}));

describe("Daftar Produk Component", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<DaftarProduk />);
  });
  it("should render with data correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should handle initial state", () => {
    expect(produkReducer(undefined, { type: "unknown" })).toEqual({
      list: [],
      loading: false,
    });
  });

  const initialState = {
    list: ["new", "value"],
    loading: false,
  };
  it("should handle reset", () => {
    const actual = produkReducer(initialState, reset());
    expect(actual.list).toEqual([]);
  });

  it("should pass", async () => {
    const postSpy = jest
      .spyOn(axios, "get")
      .mockResolvedValue({ data: { id: "1" } });
    const store = configureStore({
      reducer: function (state = "", action) {
        switch (action.type) {
          case "products/getProducts/fulfilled":
            return action.payload;
          default:
            return state;
        }
      },
    });
    await store.dispatch(getProducts());
    expect(postSpy).toBeCalledWith("listProductBySellerId?seller_id=undefined");
    const state = store.getState();
    expect(state).toEqual([]);
  });

  it("should render each elements", () => {
    expect(wrapper.find("input").props()["name"]).toBe("sellerId");
    expect(wrapper.find("form")).toHaveLength(1);
    expect(wrapper.find("button")).toHaveLength(2);
  });

  it("should submit form", () => {
    const useDispatchSpy = jest.spyOn(redux, "useDispatch");
    const mockDispatchFn = jest.fn();
    useDispatchSpy.mockReturnValue(mockDispatchFn);

    const eventSubmit = jest
      .spyOn(document, "getElementById")
      .mockImplementation(() => document.createElement("div"));
    const context = mount(
      <DaftarProduk>
        <form onSubmit={eventSubmit}>
        </form>
      </DaftarProduk>
    );
    context.find("form").simulate("submit", eventSubmit);
    expect(eventSubmit).toHaveBeenCalledTimes(2);

    useDispatchSpy.mockClear();
  });
});
