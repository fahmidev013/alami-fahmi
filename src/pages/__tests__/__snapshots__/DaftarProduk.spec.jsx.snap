// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Daftar Produk Component should render with data correctly 1`] = `
<div
  className="site-layout-content"
>
  <form
    className="inputForm"
    onSubmit={[Function]}
  >
    <input
      id="sellerId"
      name="sellerId"
      placeholder="Ketik ID Seller"
      type="text"
    />
    <button
      type="submit"
    >
      Filter
    </button>
       
    <button
      onClick={[Function]}
    >
      Reset
    </button>
  </form>
  <div
    style={
      Object {
        "maxHeight": "65vh",
        "overflow": "scroll",
      }
    }
  >
    <TableComponent
      data={
        Array [
          Object {
            "deskripsi": "Alpukat Mentega",
            "hargaSatuan": 35000,
            "id": 25,
            "key": 25,
            "nama": "Alpukat",
            "satuan": "kg",
            "sellerId": 65,
          },
          Object {
            "deskripsi": "Balsem Oles",
            "hargaSatuan": 2000,
            "id": 31,
            "key": 31,
            "nama": "Balsem",
            "satuan": "botol",
            "sellerId": 65,
          },
        ]
      }
    />
  </div>
</div>
`;
